File: 2020-11-20/Riddler Classic/riddler.py
# https://fivethirtyeight.com/features/can-you-pass-the-cranberry-sauce/

import random

def findLastToSauce(k):
    seats = list(range(k))
    wantsSauce = list(range(1, k))
    hasSauce = 0
    while(len(wantsSauce) > 1):
        passTo = random.sample([-1,1], 1)[0]
        nextSauce = (hasSauce + passTo) % k
        hasSauce = seats[nextSauce]
        if(hasSauce in wantsSauce):
            wantsSauce.remove(hasSauce)
    return wantsSauce[0]

def findLastToSauceRepeatedly(n, k):
    lastToSauce = []
    for i in range(n):
        lastToSauce.append(findLastToSauce(k))
    return [lastToSauce.count(i) / n for i in range(k)]

findLastToSauceRepeatedly(100000, 9)

File: 2020-11-20/Riddler Classic/riddler.jl
# https://fivethirtyeight.com/features/can-you-pass-the-cranberry-sauce/

using StatsBase

function findLastToSauce(k)
    seats = [0:(k-1);]
    wantsSauce = [1:(k-1);]
    hasSauce = 0
    while length(wantsSauce) > 1
        passTo = sample([-1,1])
        hasSauce = mod(hasSauce + passTo, k)
        if hasSauce in wantsSauce
            deleteat!(wantsSauce, wantsSauce .== hasSauce)
        end
    end
    return wantsSauce
end

function findLastToSauceRepeatedly(n, k)
    x = [findLastToSauce(k) for i in 1:n]
    return proportionmap(x)
end

findLastToSauceRepeatedly(100000, 20)
